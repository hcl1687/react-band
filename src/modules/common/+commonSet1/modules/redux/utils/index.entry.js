import getStore from '../getStore'

const entry = () => {
  return {
    getStore
  }
}

export default {
  entry
}
