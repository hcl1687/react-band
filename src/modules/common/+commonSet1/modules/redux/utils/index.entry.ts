import getStore from '../getStore'

export default (): RB.IRBModule => {
  return {
    getStore
  }
}
