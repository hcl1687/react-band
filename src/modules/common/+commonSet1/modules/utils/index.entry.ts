function setStatic (key: string, value: string): Utils.IComponentDeco {
  return BaseComponent => {
    BaseComponent[key] = value
    return BaseComponent
  }
}

function setDisplayName (displayName: string): Utils.IComponentDeco {
  return setStatic('displayName', displayName)
}

function getDisplayName (Component: RB.IRBComponent|string): string|undefined {
  if (typeof Component === 'string') {
    return Component
  }

  if (!Component) {
    return undefined
  }

  return Component.displayName || Component.name || 'Component'
}

function wrapDisplayName (BaseComponent: RB.IRBComponent|string, hocName: string): string {
  return `${hocName}(${getDisplayName(BaseComponent)})`
}

function promisify (fun: (...args: Array<any>) => any): ((...args: Array<any>) => Promise<any>) {
  return (...args: Array<any>) => {
    return new Promise((resolve, reject) => {
      try {
        const ret = fun(...args)
        if (ret && ret.then) {
          ret.then(resolve).catch(reject)
        } else {
          resolve(ret)
        }
      } catch (e) {
        reject(e)
      }
    })
  }
}

function getUrlParameter (url: string, name: string): string {
  // eslint-disable-next-line
  name = name.replace(/[\[]/, '\\[').replace(/[\]]/, '\\]')
  const regex = new RegExp('[\\?&]' + name + '=([^&#]*)')
  const results = regex.exec(url)
  return results === null ? '' : decodeURIComponent(results[1].replace(/\+/g, ' '))
}

const entry = (): RB.IRBModule => {
  return {
    setStatic,
    setDisplayName,
    getDisplayName,
    wrapDisplayName,
    promisify,
    getUrlParameter
  }
}

export default {
  entry
}
