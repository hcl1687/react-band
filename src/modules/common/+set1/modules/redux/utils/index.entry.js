import getStore from '../getStore'

export default () => {
  return {
    getStore
  }
}
