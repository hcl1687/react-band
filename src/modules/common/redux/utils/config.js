export default () => {
  return {
    name: 'reduxUtils'
  }
}
