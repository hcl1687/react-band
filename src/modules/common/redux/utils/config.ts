import { IRBLeafConfig } from '~/interface'

export default (): IRBLeafConfig => {
  return {
    name: 'reduxUtils'
  }
}
