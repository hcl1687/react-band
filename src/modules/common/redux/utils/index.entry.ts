import { IRBModule } from '~/interface'
import getStore from '../getStore'

export default (): IRBModule => {
  return {
    getStore
  }
}
