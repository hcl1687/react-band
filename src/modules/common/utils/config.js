export default (config) => {
  return {
    name: 'utils'
  }
}
