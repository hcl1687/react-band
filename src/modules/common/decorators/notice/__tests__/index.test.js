import React, { Children, Component, useEffect, useRef, useState } from 'react'
import PropTypes from 'prop-types'
import { mount, render } from 'enzyme'
import noticeDecoFactory from '../index.entry'
import tools from '~/../tests/utils/index'
import utils from '~/../tests/utils/mockUtils'

const context = {
  getModule: async (key) => {
    if (key === 'utils') {
      return utils
    }
  }
}

async function TestFactory () {
  class ChildTest extends Component {
    constructor (props, context) {
      super(props, context)
  
      this.state = {
        text: ''
      }
    }
  
    show1 = (text) => {
      this.setState({
        text
      })
    }
  
    show2 = (text, resolve) => {
      this.setState({
        text
      }, () => {
        resolve()
      })
    }
  
    show3 = (text) => {
      return new Promise(resolve => {
        this.setState({
          text
        }, () => {
          resolve()
        })
      })
    }
  
    render () {
      return <div id='childTest' {...this.props}>{this.state.text}</div>
    }
  }

  class Test extends Component {
    constructor (props, context) {
      super(props, context)
  
      this.state = {
        msg: ''
      }
    }

    static propTypes = {
      notify: PropTypes.func.isRequired,
      getNotification: PropTypes.func.isRequired
    }
  
    handleBtn1 = () => {
      const { notify } = this.props
      notify('child', 'show1', 'show1').then(() => {
        this.setState({
          msg: 'show1'
        })
      })
    }
  
    handleBtn2 = () => {
      const { notify } = this.props
      notify('child', 'show2', 'show2').then(() => {
        this.setState({
          msg: 'show2'
        })
      })
    }
  
    handleBtn3 = () => {
      const { notify } = this.props
      notify('child', 'show3', 'show3').then(() => {
        this.setState({
          msg: 'show3'
        })
      })
    }

    handleBtn4 = () => {
      const { notify } = this.props
      notify('child1', 'show1', 'show1').catch((err) => {
        this.setState({
          msg: err.message
        })
      })
    }

    handleBtn5 = () => {
      const { notify } = this.props
      notify('child', 'show5', 'show5').catch((err) => {
        this.setState({
          msg: err.message
        })
      })
    }
  
    render () {
      const { getNotification } = this.props
      return <div id='test' {...this.props}>
        <ChildTestWithNoticeDeco {...getNotification('child')} />
        <div id='testResult'>{this.state.msg}</div>
        <button id='btn1' onClick={this.handleBtn1}>btn1</button>
        <button id='btn2' onClick={this.handleBtn2}>btn2</button>
        <button id='btn3' onClick={this.handleBtn3}>btn3</button>
        <button id='btn4' onClick={this.handleBtn4}>btn4</button>
        <button id='btn5' onClick={this.handleBtn5}>btn5</button>
      </div>
    }
  }

  const NoticeDeco = await noticeDecoFactory(context)
  const ChildTestWithNoticeDeco = await NoticeDeco({})(ChildTest)
  const TestWithNoticeDeco = await NoticeDeco({})(Test)

  return TestWithNoticeDeco
}

async function Test1Factory () {
  const NoticeDeco = await noticeDecoFactory(context)
  const ChildTest1WithNoticeDeco = await NoticeDeco({})(ChildTest1)
  const Test1WithNoticeDeco = await NoticeDeco({})(Test1)

  function Test1 (props) {
    const { notify, getNotification } = props
    const [msg, setMsg] = useState('')

    const handleBtn1 = () => {
      notify('child', 'show1', 'show1').then(() => {
        setMsg('show1')
      })
    }

    const handleBtn2 = () => {
      notify('child', 'show2', 'show2').then(() => {
        setMsg('show2')
      })
    }

    const handleBtn3 = () => {
      notify('child', 'show3', 'show3').then(() => {
        setMsg('show3')
      })
    }

    const handleBtn4 = () => {
      notify('child1', 'show1', 'show1').catch((err) => {
        setMsg(err.message)
      })
    }

    const handleBtn5 = () => {
      notify('child', 'show5', 'show5').catch((err) => {
        setMsg(err.message)
      })
    }
  
    return <div id='test' {...props}>
      <ChildTest1WithNoticeDeco {...getNotification('child')} />
      <div id='testResult'>{msg}</div>
      <button id='btn1' onClick={handleBtn1}>btn1</button>
      <button id='btn2' onClick={handleBtn2}>btn2</button>
      <button id='btn3' onClick={handleBtn3}>btn3</button>
      <button id='btn4' onClick={handleBtn4}>btn4</button>
      <button id='btn5' onClick={handleBtn5}>btn5</button>
    </div>
  }
  
  Test1.propTypes = {
    notify: PropTypes.func.isRequried,
    getNotification: PropTypes.func.isRequried
  }
  
  function ChildTest1 (props) {
    const { setNotifyHandler } = props
    const [text, setText] = useState('')
    const thisRef = useRef({})
  
    const show1 = (text) => {
      setText(text)
    }
  
    const show2 = (text, resolve) => {
      setText(text)
      if (thisRef && thisRef.current) {
        thisRef.current.show1Resolve = resolve
      }
    }
  
    const show3 = (text) => {
      return new Promise(resolve => {
        setText(text)
        if (thisRef && thisRef.current) {
          thisRef.current.show2Resolve = resolve
        }
      })
    }
  
    setNotifyHandler({
      show1,
      show2,
      show3
    })
  
    useEffect(() => {
      if (thisRef && thisRef.current) {
        if (thisRef.current.show1Resolve) {
          const resolve = thisRef.current.show1Resolve
          delete thisRef.current.show1Resolve
          resolve()
        } else if (thisRef.current.show2Resolve) {
          const resolve = thisRef.current.show2Resolve
          delete thisRef.current.show2Resolve
          resolve()
        }
      }
    }, [text])
  
    return <div id='childTest' {...props}>{text}</div>
  }
  
  ChildTest1.propTypes = {
    setNotifyHandler: PropTypes.func.isRequried
  }

  return Test1WithNoticeDeco
}

describe('common/decorators/notice test class', () => {
  it('should render correctly', async () => {
    const TestWithNoticeDeco = await TestFactory()
    const wrapper = render(<TestWithNoticeDeco />)
    expect(wrapper).toMatchSnapshot()
  })

  // it('should render function correctly', async () => {
  //   const Test1WithNoticeDeco = await Test1Factory()
  //   const wrapper = render(<Test1WithNoticeDeco />)
  //   expect(wrapper).toMatchSnapshot()
  // })

  it('should mount correctly', async () => {
    const TestWithNoticeDeco = await TestFactory()
    const wrapper = mount(
      <TestWithNoticeDeco />
    )

    expect(typeof wrapper.find('#test').prop('getNotification')).toBe('function')
    expect(typeof wrapper.find('#test').prop('notify')).toBe('function')
    expect(typeof wrapper.find('#childTest').prop('getNotification')).toBe('function')
    expect(typeof wrapper.find('#childTest').prop('notify')).toBe('function')
    expect(wrapper.find('#childTest').prop('notification')).toEqual({})
    wrapper.find('#btn1').simulate('click')

    return tools.delay(() => {
      expect(wrapper.find('#testResult').text()).toBe('show1')
      expect(wrapper.find('#childTest').text()).toBe('show1')
    })
  })

  it('should reject notify if not register first', async () => {
    const TestWithNoticeDeco = await TestFactory()
    const wrapper = mount(
      <TestWithNoticeDeco />
    )

    wrapper.find('#btn4').simulate('click')

    return tools.delay(() => {
      expect(wrapper.find('#testResult').text()).toBe('child1 has not been registered.')
    })
  })

  it('should handle notification when called function declares a resolve parameter.', async () => {
    const TestWithNoticeDeco = await TestFactory()
    const wrapper = mount(
      <TestWithNoticeDeco />
    )

    wrapper.find('#btn2').simulate('click')

    return tools.delay(() => {
      expect(wrapper.find('#testResult').text()).toBe('show2')
      expect(wrapper.find('#childTest').text()).toBe('show2')
    })
  })

  it('should handle notification when called function return a promise', async () => {
    const TestWithNoticeDeco = await TestFactory()
    const wrapper = mount(
      <TestWithNoticeDeco />
    )

    wrapper.find('#btn3').simulate('click')

    return tools.delay(() => {
      expect(wrapper.find('#testResult').text()).toBe('show3')
      expect(wrapper.find('#childTest').text()).toBe('show3')
    })
  })

  it('should handle not existed method', async () => {
    const TestWithNoticeDeco = await TestFactory()
    const wrapper = mount(
      <TestWithNoticeDeco />
    )

    wrapper.find('#btn5').simulate('click')

    return tools.delay(() => {
      expect(wrapper.find('#testResult').text()).toBe('show5 is not exsist.')
      expect(wrapper.find('#childTest').text()).toBe('')
    })
  })
})

describe('common/decorators/notice test function', () => {
  it('should render correctly', async () => {
    const Test1WithNoticeDeco = await Test1Factory()
    const wrapper = render(<Test1WithNoticeDeco />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should mount correctly', async () => {
    const Test1WithNoticeDeco = await Test1Factory()
    const wrapper = mount(
      <Test1WithNoticeDeco />
    )

    expect(typeof wrapper.find('#test').prop('getNotification')).toBe('function')
    expect(typeof wrapper.find('#test').prop('notify')).toBe('function')
    expect(typeof wrapper.find('#childTest').prop('getNotification')).toBe('function')
    expect(typeof wrapper.find('#childTest').prop('notify')).toBe('function')
    expect(wrapper.find('#childTest').prop('notification')).toEqual({})
    wrapper.find('#btn1').simulate('click')

    return tools.delay(() => {
      expect(wrapper.find('#testResult').text()).toBe('show1')
      expect(wrapper.find('#childTest').text()).toBe('show1')
    })
  })

  it('should reject notify if not register first', async () => {
    const Test1WithNoticeDeco = await Test1Factory()
    const wrapper = mount(
      <Test1WithNoticeDeco />
    )

    wrapper.find('#btn4').simulate('click')

    return tools.delay(() => {
      expect(wrapper.find('#testResult').text()).toBe('child1 has not been registered.')
    })
  })

  it('should handle notification when called function declares a resolve parameter.', async () => {
    const Test1WithNoticeDeco = await Test1Factory()
    const wrapper = mount(
      <Test1WithNoticeDeco />
    )

    wrapper.find('#btn2').simulate('click')

    return tools.delay(() => {
      expect(wrapper.find('#testResult').text()).toBe('show2')
      expect(wrapper.find('#childTest').text()).toBe('show2')
    })
  })

  it('should handle notification when called function return a promise', async () => {
    const Test1WithNoticeDeco = await Test1Factory()
    const wrapper = mount(
      <Test1WithNoticeDeco />
    )

    wrapper.find('#btn3').simulate('click')

    return tools.delay(() => {
      expect(wrapper.find('#testResult').text()).toBe('show3')
      expect(wrapper.find('#childTest').text()).toBe('show3')
    })
  })

  it('should handle not existed method', async () => {
    const Test1WithNoticeDeco = await Test1Factory()
    const wrapper = mount(
      <Test1WithNoticeDeco />
    )

    wrapper.find('#btn5').simulate('click')

    return tools.delay(() => {
      expect(wrapper.find('#testResult').text()).toBe('show5 is not exsist.')
      expect(wrapper.find('#childTest').text()).toBe('')
    })
  })
})

describe('common/decorators/notice/production', () => {
  const NODE_ENV = process.env.NODE_ENV
  beforeAll(() => {
    process.env.NODE_ENV = 'production'
  })
  afterAll(() => {
    process.env.NODE_ENV = NODE_ENV
  })
  it('should render correctly in production env', async () => {
    const TestWithNoticeDeco = await TestFactory()
    const wrapper = render(<TestWithNoticeDeco />)
    expect(wrapper).toMatchSnapshot()
  })
})
