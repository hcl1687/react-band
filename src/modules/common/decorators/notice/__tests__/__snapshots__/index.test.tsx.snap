// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/decorators/notice test class should render correctly 1`] = `
<div
  data-props="[object Object]"
  id="test"
>
  <div
    data-props="[object Object]"
    id="childTest"
  />
  <div
    id="testResult"
  />
  <button
    id="btn1"
  >
    btn1
  </button>
  <button
    id="btn2"
  >
    btn2
  </button>
  <button
    id="btn3"
  >
    btn3
  </button>
  <button
    id="btn4"
  >
    btn4
  </button>
  <button
    id="btn5"
  >
    btn5
  </button>
</div>
`;

exports[`common/decorators/notice test function should render correctly 1`] = `
<div
  data-props="[object Object]"
  id="test"
>
  <div
    data-props="[object Object]"
    id="childTest"
  />
  <div
    id="testResult"
  />
  <button
    id="btn1"
  >
    btn1
  </button>
  <button
    id="btn2"
  >
    btn2
  </button>
  <button
    id="btn3"
  >
    btn3
  </button>
  <button
    id="btn4"
  >
    btn4
  </button>
  <button
    id="btn5"
  >
    btn5
  </button>
</div>
`;

exports[`common/decorators/notice/production should render correctly in production env 1`] = `
<div
  data-props="[object Object]"
  id="test"
>
  <div
    data-props="[object Object]"
    id="childTest"
  />
  <div
    id="testResult"
  />
  <button
    id="btn1"
  >
    btn1
  </button>
  <button
    id="btn2"
  >
    btn2
  </button>
  <button
    id="btn3"
  >
    btn3
  </button>
  <button
    id="btn4"
  >
    btn4
  </button>
  <button
    id="btn5"
  >
    btn5
  </button>
</div>
`;
